### PR TITLE
Only write cookies on the client

### LINKED ISSUE #10188: Update experiment code to deal with anonymous page caching
Because experiments rely on cookies, we need to make sure that our anonymous page caching scheme doesn't interfere with the running of experiments, and vice versa.

Some things to consider:
1. We need the correct cookie to be sent for each user in the experiment.
2. Ideally we should be able to cache a page for an experiment user (as opposed to disabling caching for anyone enrolled in an experiment).
3. We cannot `Vary` on a particular cookie value.

Some suggestions from our chat in Slack:
1. We could have a single cookie that lists all experiments a user in enrolled in, along with the corresponding variant. Then nginx can use that value as part of the cache key.
2. We should always clean up cookies after an experiment is over. Currently we just set them to expire in 30 days, but we should add code to delete them (or rather, update the single cookie mentioned above) when the experiment is no longer active.
3. If we change an experiment (i.e., change some of the UX associated with a variant) we need to ensure that we assign a new, unique id for the experiment. This is something we would generally do anyway, but it becomes critical now, because of caching.
4. We could use a scheme similar to what addons-server does for activity logs by adding all experiment ids to a file, keeping old ones around, having a list with them all and adding a test to make sure the list only contains unique items.

@diox Can you think of anything we discussed that I haven't captured?

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #10188 

POC, without tests.

### CODE REVIEW

**Comment by bobsilverberg** [comment]
@willdurand This is meant to address the issue that @diox and I were discussing on Slack. There is a problem with setting a cookie on the server, and our hope is that an approach like this might work where we defer setting cookies to the client, in a similar way that we are deferring sending events to GA to the client.

I think with this code we can probably also get rid of `react-cookie` and `withCookies`, because we don't actually need them, and can just use `universal-cookie` (or some other library if a client-only, developer-friendly one exists).

Could you please take a look at this draft PR when you have some time. I don't want to start writing tests until I know whether this seems like a viable route to take.

**Comment by willdurand** [comment]
> I think with this code we can probably also get rid of react-cookie and withCookies, because we don't actually need them, and can just use universal-cookie (or some other library if a client-only, developer-friendly one exists).

How do we authenticate users then?

**Comment by willdurand** [comment]
> There is a problem with setting a cookie on the server, and our hope is that an approach like this might work where we defer setting cookies to the client, in a similar way that we are deferring sending events to GA to the client

Could we move the whole enrollment logic to the client side then? (I didn't look at the code yet)

**Comment by bobsilverberg** [comment]
> > I think with this code we can probably also get rid of react-cookie and withCookies, because we don't actually need them, and can just use universal-cookie (or some other library if a client-only, developer-friendly one exists).
> 
> How do we authenticate users then?

This shouldn't affect authentication. It's inaccurate of me to say that this PR changes the way we deal with *all* cookies. This only affects the cookie for experiments, and provides a way, moving forward, for us to ensure that cookies are only set on the client, when that is important.

As this is being done to deal with a caching issue, and we don't cache pages for authenticated users, we shouldn't have to change anything about our authentication cookie.

**Comment by bobsilverberg** [comment]
> > There is a problem with setting a cookie on the server, and our hope is that an approach like this might work where we defer setting cookies to the client, in a similar way that we are deferring sending events to GA to the client
> 
> Could we move the whole enrollment logic to the client side then? (I didn't look at the code yet)

It's possible, but I think the way I encapsulated the cookie-setting logic would make that more difficult. Basically, we call a function to set a cookie when the cookie needs to be set, and that's the point at which we also do the enrollment event. That can happen on either the client or the server, and the code that sets the cookie, and sends the event, doesn't need to care about whether it is running on the client or server. This means that sometimes these things are initiated on the client, and sometimes on the server, and everything is designed so that the calling code doesn't need to care about whether it is on the client or server.

**Comment by willdurand** [comment]
> It's possible, but I think the way I encapsulated the cookie-setting logic would make that more difficult. Basically, we call a function to set a cookie when the cookie needs to be set, and that's the point at which we also do the enrollment event. That can happen on either the client or the server, and the code that sets the cookie, and sends the event, doesn't need to care about whether it is running on the client or server. This means that sometimes these things are initiated on the client, and sometimes on the server, and everything is designed so that the calling code doesn't need to care about whether it is on the client or server.

Maybe we should handle all of this on the client _only_ then? I didn't really like the changes to transfer GA events from the server to the client, and seeing the same code again but for cookies makes me nervous.

Could we read the cookie to determine the variant on the client and server and, if the variant doesn't exist:

- on the server: we assign a variant, put it in the redux store. The client would read that value and create the cookie + ping GA
- on the client: is it even possible? We don't remote control experiments. If we do want to support this, then the code to assign a variant would have to work no matter the context.

In other words, reading cookies could happen in any context, assigning a variant (`Math.random()`) too, but _writing_ a cookie should be a pure client side thing. If I understand correctly, the enrollment ping to GA is needed when we write a cookie after having put the user in a bucket, so that would be a client-side thing, too, which would make sense. So I am thinking that we might simplify a few things here, _if_ we want to use that approach.
